File: src/ObjectManager/Object.Tes/FilePacks/Records/300-BODY.Body.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class BODYRecord : Record
    {
        public struct BYDTField
        {
            public byte Part;
            public byte Vampire;
            public byte Flags;
            public byte PartType;

            public BYDTField(UnityBinaryReader r, int dataSize)
            {
                Part = r.ReadByte();
                Vampire = r.ReadByte();
                Flags = r.ReadByte();
                PartType = r.ReadByte();
            }
        }

        public override string ToString() => $"BODY: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // NIF Model
        public STRVField FNAM; // Body name
        public BYDTField BYDT;

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            if (format == GameFormatId.TES3)
                switch (type)
                {
                    case "NAME": EDID = new STRVField(r, dataSize); return true;
                    case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                    case "FNAM": FNAM = new STRVField(r, dataSize); return true;
                    case "BYDT": BYDT = new BYDTField(r, dataSize); return true;
                    default: return false;
                }
            return false;
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/300-BODY.Body.swift
//
//  BODYRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class BODYRecord: Record {
    public struct BYDTField {
        public let part: UInt8
        public let vampire: UInt8
        public let flags: UInt8
        public let partType: UInt8

        init(_ r: BinaryReader, _ dataSize: Int) {
            part = r.readByte()
            vampire = r.readByte()
            flags = r.readByte()
            partType = r.readByte()
        }
    }

    public override var description: String { return "BODY: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var MODL: MODLGroup? = nil // NIF Model
    public var FNAM: STRVField! // Body name
    public var BYDT: BYDTField!

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        guard format == .TES3 else {
            return false
        }
        switch type {
        case "NAME": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "FNAM": FNAM = r.readSTRV(dataSize)
        case "BYDT": BYDT = BYDTField(r, dataSize)
        default: return false
        }
        return true
    }
}
